server/app/controllers/applicant/RedirectController.java

```diff
@@ -107,32 +111,39 @@ public CompletionStage<Result> programBySlug(Http.Request request, String progra
                         // with that application if so.
                         if (programForExistingApplication.isPresent()) {
+                          long programId = programForExistingApplication.get().id();
                           return CompletableFuture.completedFuture(
-                              redirect(
-                                      controllers.applicant.routes.ApplicantProgramReviewController
-                                          .review(
-                                              applicantId,
-                                              programForExistingApplication.get().id()))
-                                  .flashing("redirected-from-program-slug", programSlug));
+                              redirectToReviewPage(programId, applicantId, programSlug, request));
+                        } else {
+                          return programService
+                              .getActiveProgramDefinitionAsync(programSlug)
+                              .thenApply(
+                                  activeProgramDefinition ->
+                                      redirectToReviewPage(
+                                          activeProgramDefinition.id(),
+                                          applicantId,
+                                          programSlug,
+                                          request))
+                              .exceptionally(
+                                  ex ->
+                                      notFound(ex.getMessage())
+                                          .removingFromSession(request, REDIRECT_TO_SESSION_KEY));
                         }
-
-                        return redirectToActiveProgram(applicantId, programSlug);
                       },
                       httpContext.current());
             },
             httpContext.current());
   }
 
-  private CompletionStage<Result> redirectToActiveProgram(long applicantId, String programSlug) {
-    return programService
-        .getActiveProgramDefinitionAsync(programSlug)
-        .thenApplyAsync(
-            (activeProgramDefinition) ->
-                redirect(
-                        controllers.applicant.routes.ApplicantProgramReviewController.review(
-                            applicantId, activeProgramDefinition.id()))
-                    .flashing("redirected-from-program-slug", programSlug),
-            httpContext.current());
+  private Result redirectToReviewPage(
+      long programId, long applicantId, String programSlug, Http.Request request) {
+    return redirect(
+            controllers.applicant.routes.ApplicantProgramReviewController.review(
+                applicantId, programId))
+        .flashing("redirected-from-program-slug", programSlug)
+        // If we had a redirectTo session key that redirected us here, remove it so that it doesn't
+        // get used again.
+        .removingFromSession(request, REDIRECT_TO_SESSION_KEY);
```

Comment:
**optional:** I think you could revert my patch from earlier now that this is in place.

Reply:
I think we're better off keeping it since it's better practice to never have a `redirectTo` when landing on the index page.

Ideally we would avoid redirects in `LogoutController`, but the version of that class that we're using is Play's, and it doesn't seem worth it to try to mess with it.